File: _C/src/python/bindings/linalg.cc
#include "jetdl/linalg.h"

#include <pybind11/pybind11.h>

#include "jetdl/python/linalg/bindings.h"

namespace py = pybind11;

void bind_linalg_submodule(py::module_& m) {
  py::module_ linalg = m.def_submodule("linalg");

  linalg.def("c_dot", &jetdl::linalg::dot);
  linalg.def("c_matmul", &jetdl::linalg::matmul);

  linalg.def("c_T", &jetdl::linalg::T);
  linalg.def("c_mT", &jetdl::linalg::mT);
}

File: jetdl/linalg.py
from ._C import Tensor
from ._C.linalg import c_dot, c_matmul, c_mT, c_T


def dot(a: Tensor, b: Tensor) -> Tensor:
    return c_dot(a, b)


def matmul(a: Tensor, b: Tensor) -> Tensor:
    return c_matmul(a, b)


def transpose(input: Tensor) -> Tensor:
    return c_T(input)


def matrix_transpose(input: Tensor) -> Tensor:
    return c_mT(input)
